### DISCUSSION TITLE
Increase MaxModuleNestingDepth

### DISCUSSION THREAD

**Discussion by notatechproblem** [PowerShell/PowerShell#19777] [Enhancements]
I make extensive reuse of custom classes in my PowerShell code, so I've started using script modules as class libraries that I reference with `using module MyModule` statements. However, I've had to start grouping classes together into fewer, larger modules because having my class library modules call other class library modules very quickly results in:

`Cannot load the module 'MyModule.psd1' because the module nesting limit has been exceeded`

It looks like the limit is hard-coded in ModuleIntrinsics.cs:

`private const int MaxModuleNestingDepth = 10;`

My C# skills aren't great, so I'm having a difficult time figuring out if that limit is based on technical reasons, or if it was just pulled out of the air by some developer on the PowerShell team thinking "No one will ever need to nest modules more than 10 deep!". According to blame, that line of code has been unchanged since the first commit to the PS repo in 2016.

Being able to break up my classes into smaller class library modules would be a significant improvement to my devEx, and at first glance it seems like it would be very easy to fix.

Does anyone know if there are any technical reasons preventing MaxModuleNestingDepth from being set to a number larger than 10? Or, could the value be stored instead as a customizable value that is read from powershell.config.json instead of being hard-coded?

**Comment by splatteredbits** [comment]
I'm having this problem as well. In practice, it turns out the actual depth is half of 10, because when you use `Import-Module` it imports the .psd1 file which then imports the .psm1 file. Each of those imports counts as a nesting context, so you can only nest to five modules deep.

We make extensive use of nested modules so that our modules can ship with their own dependencies so users don't have to track, manage, and import them. A module can be its own self-contained unit. And we've just hit this 10 nesting limit, er, five nested limit.

This should change and should probably be increased by an order of magnitude. Unfortunately, this doesn't help those of us still extensively using PowerShell 5.1.

**Comment by splatteredbits** [comment]
As a workaround, import the .psm1 file instead of the module by name or by .psd1. So, instead of:

```
Import-Module -Name (Join-Path -Path $PSScriptRoot -ChildPath 'Modules\Carbon' -Resolve)
```

you can

```
Import-Module -Name (Join-Path -Path $PSScriptRoot -ChildPath 'Modules\Carbon\Carbon.psm1' -Resolve)
```

The module you import has to be written so that it doesn't use the .psd1 file to auto-load anything you'll need, like assemblies, formats, type data, etc.,

**Comment by splatteredbits** [comment]
More information for anyone that needs to workaround this: the nesting limit only seems to be an issue when you first import a module. So if you have a module that is deeply nested, try importing it at a higher nesting level before it gets imported at a deeper nesting level.
